flink-kubernetes/src/test/java/org/apache/flink/kubernetes/highavailability/KubernetesJobGraphStoreWatcherTest.java

```diff
@@ -0,0 +1,135 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.kubernetes.highavailability;
+
+import org.apache.flink.api.common.JobID;
+import org.apache.flink.kubernetes.kubeclient.FlinkKubeClient;
+import org.apache.flink.kubernetes.kubeclient.resources.KubernetesConfigMap;
+import org.apache.flink.kubernetes.utils.KubernetesUtils;
+import org.apache.flink.runtime.jobmanager.JobGraphStoreWatcher;
+import org.apache.flink.runtime.jobmanager.TestingJobGraphListener;
+
+import org.junit.Before;
+import org.junit.Test;
+
+import java.util.Collections;
+import java.util.concurrent.TimeUnit;
+
+import static org.hamcrest.Matchers.contains;
+import static org.hamcrest.Matchers.containsInAnyOrder;
+import static org.hamcrest.Matchers.is;
+import static org.junit.Assert.assertThat;
+
+/**
+ * Tests for {@link KubernetesRunningJobsRegistry} operations.
+ */
+public class KubernetesJobGraphStoreWatcherTest extends KubernetesHighAvailabilityTestBase {
+
+	private TestingJobGraphListener testingJobGraphListener;
+
+	@Before
+	public void setup() {
+		super.setup();
+		testingJobGraphListener = new TestingJobGraphListener();
+	}
+
+	@Test
+	public void testJobGraphAddedAndRemovedShouldNotifyJobGraphListener() throws Exception {
+		new Context() {{
+			runTest(
+				() -> {
+					leaderCallbackGrantLeadership();
+
+					final JobGraphStoreWatcher jobGraphStoreWatcher =
+						createAndStartJobGraphStoreWatcher(flinkKubeClient);
+
+					final JobID jobID = JobID.generate();
+					final String key = KubernetesUtils.jobIDToJobGraphConfigMapKey(jobID);
+
+					assertThat(configMapCallbackFutures.size(), is(3));
+					final FlinkKubeClient.WatchCallbackHandler<KubernetesConfigMap> callbackHandler =
+						configMapCallbackFutures.get(2).get(TIMEOUT, TimeUnit.MILLISECONDS);
+					// A new job graph is added
+					getLeaderConfigMap().getData().put(key, "job graph data");
+					callbackHandler.onModified(Collections.singletonList(getLeaderConfigMap()));
+
+					assertThat(testingJobGraphListener.getAddedJobGraphs(), contains(jobID));
+
+					// Job graph is deleted
+					getLeaderConfigMap().getData().remove(key);
+					callbackHandler.onModified(Collections.singletonList(getLeaderConfigMap()));
+
+					assertThat(testingJobGraphListener.getRemovedJobGraphs(), contains(jobID));
+
+					jobGraphStoreWatcher.stop();
+				});
+		}};
+	}
+
+	/**
+	 * We already have three job graphs in the store, job0 and job1 is removed and job3 is newly added.
+	 */
+	@Test
+	public void testMultipleJobGraphChanges() throws Exception {
```

Comment:
I think this test covers the `testJobGraphAddedAndRemovedShouldNotifyJobGraphListener`. We may remain one of these.

Reply:
Make sense.